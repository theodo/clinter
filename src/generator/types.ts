import { Linter } from "eslint";
import { ProjectInfoObject } from "types";

export type ESLintGenerator = (userAnswers: ProjectInfoObject) => (config: Linter.Config) => Linter.Config;

export type ESLintDependencyGenerator = (userAnswers: ProjectInfoObject) => (deps: string[]) => string[];

export interface ESLintOverrider {
  files: string[];
}
