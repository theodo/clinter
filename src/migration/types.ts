export interface MigrationResults {
  fixableErrorCount: number;
  errorCount: number;
}
