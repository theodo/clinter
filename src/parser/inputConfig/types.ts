import { AnswerObject, ModeAnswerObject } from "types";

export interface InputConfig {
  generatorConfig: AnswerObject;
  modeConfig: ModeAnswerObject;
}
