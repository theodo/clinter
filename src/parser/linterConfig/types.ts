import { Options } from "prettier";
import { Linter } from "eslint";

export enum FileExtension {
  JSON = "json",
  YAML = "yml",
  JS = "js",
  NONE = "none",
}

export interface ConfigFileObject {
  name: string;
  extension: FileExtension;
  attribute?: string;
}

export type LinterConfigs = Options | Linter.Config;

export interface ConfigContainer<T extends LinterConfigs> {
  config: T;
  file: ConfigFileObject;
}
