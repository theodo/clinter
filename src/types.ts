export enum FormatterInfo {
  None = "None",
  Prettier = "Prettier",
}

export interface FormatterInfoObject {
  formatter: FormatterInfo;
}

export enum FrontFrameworkInfo {
  None = "None",
  Vue = "Vue",
  React = "React",
}

export interface FrontFrameworkInfoObject {
  frontFramework: FrontFrameworkInfo;
}

export enum TypescriptInfo {
  None = "None",
  WithTypeChecking = "With type checking",
  NoTypeChecking = "Without type checking",
}

export interface TypescriptInfoObject {
  typescript: TypescriptInfo;
}

export enum EnvInfo {
  Browser = "Browser",
  Node = "Node",
}

export interface EnvInfoObject {
  env: EnvInfo[];
}

export enum TestFrameworkInfo {
  None = "None",
  Jest = "Jest",
}

export interface TestFrameworkInfoObject {
  test: TestFrameworkInfo;
}

export type ProjectInfoObject = FormatterInfoObject &
  FrontFrameworkInfoObject &
  TypescriptInfoObject &
  EnvInfoObject &
  TestFrameworkInfoObject;

export enum ClinterModeInfo {
  Generator = "Generator Mode",
  Upgrade = "Upgrade Mode",
}

export interface ClinterModeInfoObject {
  mode: ClinterModeInfo;
}

export interface MigrationModeInfoObject {
  migration: boolean;
}

export interface DependenciesConfigObject {
  upgradeDependencies: boolean;
}

export interface ClinterSettings {
  generatorConfig: ProjectInfoObject;
  modeConfig: ClinterModeInfoObject;
  migrationModeConfig: MigrationModeInfoObject;
  dependenciesConfig: DependenciesConfigObject;
}
