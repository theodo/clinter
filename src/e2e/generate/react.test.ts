import { defaultInputConfig, makeTestService } from "e2e/helpers";
import { ClinterSettings, FrontFrameworkInfo, TypescriptInfo } from "types";

describe("React configuration Generator Mode", () => {
  const testService = makeTestService();

  it("should return a react eslint configuration", async () => {
    const inputConfig: ClinterSettings = {
      ...defaultInputConfig,
      generatorConfig: {
        ...defaultInputConfig.generatorConfig,
        frontFramework: FrontFrameworkInfo.React,
      },
    };

    testService.loadInputConfig(inputConfig);
    await testService.runClinter();
    const outputConfig = testService.getParsedOutputConfig();

    expect(outputConfig).toMatchSnapshot("eslint-react-config");
  });

  it("should return a react typescript eslint configuration with type checking", async () => {
    const inputConfig: ClinterSettings = {
      ...defaultInputConfig,
      generatorConfig: {
        ...defaultInputConfig.generatorConfig,
        typescript: TypescriptInfo.WithTypeChecking,
        frontFramework: FrontFrameworkInfo.React,
      },
    };

    const tsConfig = {
      compilerOptions: {
        strict: true,
      },
    };

    testService.loadInputConfig(inputConfig);
    testService.loadInitialTSConfig(tsConfig);

    await testService.runClinter();
    const outputConfig = testService.getParsedOutputConfig();

    expect(outputConfig).toMatchSnapshot("eslint-react-ts-config");
  });
});
