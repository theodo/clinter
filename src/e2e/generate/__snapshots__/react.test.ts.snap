// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React configuration Generator Mode should return a react eslint configuration: eslint-react-config 1`] = `
Object {
  "extends": Array [
    "eslint:recommended",
    "react-app",
    "plugin:jsx-a11y/recommended",
  ],
  "plugins": Array [
    "import",
    "jsx-a11y",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "padding-line-between-statements": Array [
      "error",
      Object {
        "blankLine": "always",
        "next": "return",
        "prev": "*",
      },
    ],
    "prefer-const": "error",
    "react-hooks/exhaustive-deps": "warn",
    "react-hooks/rules-of-hooks": "error",
    "react/no-string-refs": "warn",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
  "settings": Object {
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`React configuration Generator Mode should return a react typescript eslint configuration with type checking: eslint-react-ts-config 1`] = `
Object {
  "extends": Array [
    "eslint:recommended",
    "react-app",
    "plugin:jsx-a11y/recommended",
  ],
  "overrides": Array [
    Object {
      "extends": Array [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      "files": Array [
        "**/*.ts?(x)",
      ],
      "parser": "@typescript-eslint/parser",
      "parserOptions": Object {
        "project": "tsconfig.json",
      },
      "rules": Object {
        "@typescript-eslint/no-shadow": "error",
        "@typescript-eslint/prefer-nullish-coalescing": "error",
        "@typescript-eslint/prefer-optional-chain": "error",
        "@typescript-eslint/strict-boolean-expressions": "error",
        "no-shadow": "off",
      },
    },
  ],
  "plugins": Array [
    "import",
    "jsx-a11y",
  ],
  "root": true,
  "rules": Object {
    "@typescript-eslint/explicit-function-return-type": "off",
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "padding-line-between-statements": Array [
      "error",
      Object {
        "blankLine": "always",
        "next": "return",
        "prev": "*",
      },
    ],
    "prefer-const": "error",
    "react-hooks/exhaustive-deps": "warn",
    "react-hooks/rules-of-hooks": "error",
    "react/no-string-refs": "warn",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
  "settings": Object {
    "react": Object {
      "version": "detect",
    },
  },
}
`;
